### PR TITLE
got rid of intermediate files

### PR DESCRIPTION
**Pull request by imyiningliu**

### CODE REVIEW

**Comment by jeff-regier** [comment]
It passes for me locally too, but it needs ~ 6GB of RAM. That's too much for travis I think. Probably subsampling the genes is the way to go.

**Comment by imyiningliu** [comment]
Thanks Jeff! I tried to subsample using brain_large.py as a reference. I have a question about subsampling: 
I think we get the matrix by reading the csv file and take the first 600 rows with largest variance. Does sampling refer to only use a subset of the rows to make the original matrix? For example, we still read the csv file, but only store ~20% of the rows in the matrix? 

(I can work on this after final/when i come back during summer if it's easier to explain in person)

**Comment by jeff-regier** [comment]
I'm not entirely sure what the best way to do subsampling is: by cell or by gene. And if by gene, then how do we select the genes to keep? At random? Or take the most often observed genes? It'd be nice in the unit tests to subsample the datasets the same way they will be subsampled in practice.

I think @Edouard360 and @maxime1310  are working on a related issue: their unit test for the retina dataset apparently uses a lot of memory too. Could you try working out the issue with them?

If subsampling is something we'll do for a lot of datasets, maybe it could be in the base class (`GeneExpressionDataset`) rather than duplicated in the child classes.

File: scvi/dataset/cbmc.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        with gzip.open(self.save_path + self.download_name, "rt", encoding="utf8") as csvfile:
+            data_reader = csv.reader(csvfile, delimiter=',')
+            for i, row in enumerate(data_reader):
+                rows.append(row[1:])
+                gene_names.append(row[0])
+
+        expression_data = np.array(rows[1:], dtype=np.int).T
+        gene_names = np.array(gene_names[1:], dtype=np.str)
+
+        selected = np.std(expression_data, axis=0).argsort()[-600:][::-1]
```
Comment (jeff-regier) [inline scvi/dataset/cbmc.py]: Are we subsampling 600 of something here? I think we just want to subsample during the unit test -- not always.

**Comment by jeff-regier** [comment]
Looks good to me (except I'm not sure about the `600` above). Once the unit tests are passing, @Edouard360 will you review that change too and "Squash and merge" the pull request?

**Comment by jeff-regier** [comment]
@imyiningliu Can you make the unit test file any smaller? About 50 KB would be good. Looks like it's about 700 KB now.

**Comment by jeff-regier** [comment]
Do you know why `subsample_barcodes` isn't called by the unit test? (according to the code coverage statistics above)

**Comment by imyiningliu** [comment]
I subsampled by keeping the first 1% of barcodes: each gene has 8617 barcodes and i kept the first 86 barcodes. 
The matrix would be mostly zeros if i subsample less barcodes. Currently the dataset has 36280 
 genes. Should I subsample 10% of the genes (keeping the first 10%)? 

I commented out the line calling subsample_barcodes after getting the unit test file. Since the method probably won't be called again given that the unit test file exists, I can remove that method or put it somewhere else.

**Comment by imyiningliu** [comment]
I think it would be better if i can subsample from genes (but i'm not sure if that makes sense in a biology point of view). The matrix has very few ones now since i only kept 1% of the barcodes :(

**Comment by jeff-regier** [comment]
For the test data, you can subsample cells or genes or both to generate a small file. The important thing is just that it has the same general format as the real file, and that you can test all your code with it.

**Comment by Edouard360** [comment]
@jeff-regier 
- Is it possible to squash even when it has already been pushed ? Or you meant we should push -f the locally squashed commits ?
- Indeed, refering to your older comment in this thread we should factorize subsampling cells and genes, in the parent GeneDataset class.

**Comment by jeff-regier** [comment]
I just meant to use the "Squash and merge" button in Github once you've reviewed this PR and you're happy with it. That way all these commits will be grouped together in our git history. (We should probably be doing that for all pull requests going forward, not just this one.)

**Comment by jeff-regier** [comment]
Yes, that makes sense. Let's have both the cell and the gene (row and column) subsampling in `GeneDataset`. It's functionality that's useful for any dataset, not just this one and brain_large.

File: scvi/dataset/cbmc.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        else:
+            self.download_name = "../tests/data/cbmc_subsampled.csv.gz"
+
+        expression_data, gene_names = self.download_and_preprocess()
+
+        super(CbmcDataset, self).__init__(
+            *GeneExpressionDataset.get_attributes_from_matrix(
+                expression_data), gene_names=gene_names)
+
+    def export_unit_test(self, n_cells=50, n_genes=10):
```
Comment (jeff-regier) [inline scvi/dataset/cbmc.py]: Unit test code should go in the `tests` directory.

Comment (jeff-regier) [reply scvi/dataset/cbmc.py]: Oh, I see. Better not to include the code that creates data for unit tests at all in the repo.

File: scvi/dataset/cbmc.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        self.subsample_cells(n_cells)
```
Comment (jeff-regier) [inline scvi/dataset/cbmc.py]: Better to exclude code the generates data for unit tests from the repo.

Comment (jeff-regier) [reply scvi/dataset/cbmc.py]: Actually, I'm not sure I understand what's going on here. If this code is only for testing, can we put it in the `tests` directory, along with the unit test?

File: scvi/dataset/cbmc.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        self.subsample_genes(n_genes)
+        subsampled_data = np.concatenate((self.gene_names.reshape(1, -1), self.X)).T
+
+        with gzip.open('tests/data/cbmc_subsampled.csv.gz', 'w') as gzipfile:
```
Comment (jeff-regier) [inline scvi/dataset/cbmc.py]: I'm pretty sure we don't want the main code to have references to data in the `tests` directory anyway.

**Comment by Edouard360** [comment]
Okay so we never keep any trace of the subsampling process for unit_testing.

**Comment by jeff-regier** [comment]
Yep, that's the idea. Better not to keep code in the repo that is only used for generating test data.

**Comment by imyiningliu** [comment]
Romain used .csv in scVI-reproducibility. Right now I used .csv.gzip. 

Should I change the code to deal with .csv files instead?

**Comment by jeff-regier** [comment]
No, what you have seems fine. Let's merge this once Eddie has reviewed the changes.

**Pull request merged**